code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,200 @@ RSF
-		to_chat(user, "<span class='notice'>The RSF now holds [matter]/30 fabrication-units.</span>")
-		icon_state = "rsf"
+		to_chat(user, "<span class='notice'>\The [src] now holds [matter]/[max_matter] [discriptor].</span>")
+		icon_state = initial(icon_state)//and set the icon state to the initial value it had
 	else
 		return ..()
 
 /obj/item/rsf/attack_self(mob/user)
+	if(!user)
+		return
 	playsound(src.loc, 'sound/effects/pop.ogg', 50, FALSE)
-	switch(mode)
-		if(5)
-			mode = 1
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Drinking Glass'.</span>")
-		if(1)
-			mode = 2
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Paper'.</span>")
-		if(2)
-			mode = 3
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Pen'.</span>")
-		if(3)
-			mode = 4
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Dice Pack'.</span>")
-		if(4)
-			mode = 5
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Cigarette'.</span>")
+	var/list/item_list
+	var/atom/target = cost_by_item
+	var/cost = 0
+	//Warning, prepare for bodgecode
+	while(istype(target, /list/))//While target is a list we continue the loop
+		item_list = list()
+		for(var/meme in target)//We iterate through all of targets entrys
+			var/atom/test
+			if(istype(meme, /list/))//If it's a list let's use its associated object
+				var/temp = target[meme]
+				test = new temp()
+			else
+				test = new meme()
+			//We then add their data into the radial menu
+			item_list[test.name] = image(icon = test.icon, icon_state = test.icon_state)
+			qdel(test)
+		var/picked = show_radial_menu(user, src, item_list, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE)
+		if(!check_menu(user) || picked == null)
+			return
+		for(var/emem in target)//Back through target agian
+			var/atom/test
+			if(istype(emem, /list/))
+				var/temp = target[emem]//If it's a list we should use its associated object
+				test = new temp()
+			else
+				test = new emem()
+			if(picked == test.name)//We try and find the entry that matches the radial selection
+				cost = target[emem]//We cash the cost
+				target = emem
+				break
+			qdel(test)
+		//If we found a list we start it all again, this time looking through its contents.
+	to_despense = target
+	despense_cost = cost
 	// Change mode
 
+/obj/item/rsf/proc/check_menu(mob/user)
+	if(!istype(user))
+		return FALSE
+	if(user.incapacitated() || !user.Adjacent(src))
+		return FALSE
+	return TRUE
+
 /obj/item/rsf/afterattack(atom/A, mob/user, proximity)
 	. = ..()
 	if(!proximity)
 		return
-	if (!(istype(A, /obj/structure/table) || isfloorturf(A)))
+	if(!is_allowed(A))
 		return
+	if(use_matter(despense_cost, user))//If we can charge that amount of charge, we do so and return true
+		playsound(src.loc, 'sound/machines/click.ogg', 10, TRUE)
+		var/atom/meme = new to_despense(get_turf(A))
+		to_chat(user, "<span class='notice'>[action_type] [meme.name]...</span>")
 
```

Comment:
I think here need some like 
```suggestion
	else
		to_chat(user, "<span class='warning'>Not enough [discriptor]!</span>")

```

Reply:
use_matter() handles the text cues, maybe I should move that...